File: apps/client/lib/domain/spec_features.dart
/// Wire helpers for Element-parity MSCs. Keep this file free of Flutter widgets.
library;

const msc4139PromptsKey = 'org.matrix.msc4139.prompts';
const msc4139UsedPromptKey = 'org.matrix.msc4139.used_prompt';
const msc4139ReplyType = 'org.matrix.msc4139.conversation.reply';
const msc4332CommandsState = 'org.matrix.msc4332.commands';
const msc4310Decline = 'm.rtc.decline';
const msc4310DeclineUnstable = 'org.matrix.msc4310.rtc.decline';
const msc2545PackState = 'im.ponies.room_emotes';
const msc2545UserEmotes = 'im.ponies.user_emotes';
const stickerEventType = 'm.sticker';
const msc3266SummaryPath = '/_matrix/client/v1/room_summary';
const msc3266SummaryUnstablePath =
    '/_matrix/client/unstable/im.nheko.summary/summary';
const simplifiedSlidingSyncPath =
    '/_matrix/client/unstable/org.matrix.simplified_msc3575/sync';
const msc3575SlidingSyncPath = '/_matrix/client/unstable/org.matrix.msc3575/sync';
const msc4039Upload = 'org.matrix.msc4039.upload_file';
const msc4039Download = 'org.matrix.msc4039.download_file';

class IntentionalMentions {
  const IntentionalMentions({required this.userIds, this.room = false});

  final List<String> userIds;
  final bool room;

  Map<String, Object?> toJson() => {
        'user_ids': userIds,
        if (room) 'room': true,
      };
}

IntentionalMentions intentionalMentions(String body, Iterable<String> memberIds) {
  final ids = <String>{
    for (final id in memberIds)
      if (id.isNotEmpty && body.contains(id)) id,
  };
  final room = RegExp(r'(^|\s)@room\b', caseSensitive: false).hasMatch(body);
  return IntentionalMentions(userIds: ids.toList(), room: room);
}

Map<String, Object?> attachMentions(
  Map<String, Object?> content,
  String body,
  Iterable<String> memberIds,
) {
  return {...content, 'm.mentions': intentionalMentions(body, memberIds).toJson()};
}

Map<String, Object?> threadRelation(
  String rootId, {
  String? replyToId,
  bool? fallback,
}) {
  final reply = replyToId ?? rootId;
  return {
    'rel_type': 'm.thread',
    'event_id': rootId,
    'is_falling_back': fallback ?? reply == rootId,
    'm.in_reply_to': {'event_id': reply},
  };
}

String? threadRootId(Map<String, dynamic> content) {
  final rel = content['m.relates_to'];
  if (rel is! Map) return null;
  if (rel['rel_type'] != 'm.thread') return null;
  final id = rel['event_id'];
  return id is String ? id : null;
}

bool isThreadFallback(Map<String, dynamic> content) {
  final rel = content['m.relates_to'];
  return rel is Map && rel['is_falling_back'] == true;
}

bool belongsOnMainTimeline(Map<String, dynamic> content) {
  final root = threadRootId(content);
  if (root == null) return true;
  return isThreadFallback(content);
}

class GeoPoint {
  const GeoPoint({
    required this.lat,
    required this.lon,
    required this.geoUri,
    this.description,
  });

  final double lat;
  final double lon;
  final String geoUri;
  final String? description;
}

GeoPoint? parseGeoUri(String value) {
  final match = RegExp(r'^geo:(-?\d+(?:\.\d+)?),(-?\d+(?:\.\d+)?)(?:;|$)', caseSensitive: false)
      .firstMatch(value.trim());
  if (match == null) return null;
  final lat = double.tryParse(match.group(1)!);
  final lon = double.tryParse(match.group(2)!);
  if (lat == null || lon == null) return null;
  return GeoPoint(lat: lat, lon: lon, geoUri: value.trim());
}

Map<String, Object?> locationContent(double lat, double lon, {String? description}) {
  final geoUri = 'geo:$lat,$lon';
  final body = (description?.trim().isNotEmpty ?? false) ? description!.trim() : geoUri;
  return {
    'msgtype': 'm.location',
    'body': body,
    'geo_uri': geoUri,
    'org.matrix.msc3488.location': {
      'uri': geoUri,
      if (description != null && description.trim().isNotEmpty) 'description': description.trim(),
    },
    'org.matrix.msc3488.ts': DateTime.now().millisecondsSinceEpoch,
    'org.matrix.msc1767.text': body,
  };
}

GeoPoint? parseLocationContent(Map<String, dynamic> content) {
  final nested = content['org.matrix.msc3488.location'];
  final uri = nested is Map ? nested['uri'] : content['geo_uri'];
  if (uri is! String) return null;
  final parsed = parseGeoUri(uri);
  if (parsed == null) return null;
  final description = nested is Map ? nested['description'] : content['body'];
  return GeoPoint(
    lat: parsed.lat,
    lon: parsed.lon,
    geoUri: uri,
    description: description is String ? description : parsed.description,
  );
}

String openStreetMapUrl(double lat, double lon) =>
    'https://www.openstreetmap.org/?mlat=$lat&mlon=$lon#map=16/$lat/$lon';

class ImagePackItem {
  const ImagePackItem({
    required this.shortcode,
    required this.url,
    required this.body,
    this.usage = const ['sticker'],
  });

  final String shortcode;
  final String url;
  final String body;
  final List<String> usage;
}

List<ImagePackItem> parseImagePack(Map<String, dynamic> content) {
  final images = content['images'];
  if (images is! Map) return const [];
  final out = <ImagePackItem>[];
  images.forEach((key, raw) {
    if (raw is! Map) return;
    final url = raw['url'];
    if (url is! String || !url.startsWith('mxc://')) return;
    final usage = (raw['usage'] as List?)?.whereType<String>().toList() ??
        const ['sticker', 'emoticon'];
    out.add(
      ImagePackItem(
        shortcode: key.toString(),
        url: url,
        body: raw['body'] as String? ?? key.toString(),
        usage: usage,
      ),
    );
  });
  return out;
}

Map<String, Object?> stickerContent(ImagePackItem item) => {
      'body': item.body,
      'url': item.url,
      'info': {'mimetype': 'image/png'},
    };

class Msc4139Prompt {
  const Msc4139Prompt({
    required this.id,
    required this.label,
    this.type = 'preset',
    this.validator,
  });

  final String id;
  final String label;
  final String type;
  final String? validator;
}

class Msc4139Prompts {
  const Msc4139Prompts({
    required this.prompts,
    this.intro,
    this.scope,
  });

  final List<Msc4139Prompt> prompts;
  final String? intro;
  final List<String>? scope;
}

String _promptLabel(Object? label) {
  if (label is String) return label;
  if (label is! Map) return '';
  final text = label['m.text'] ?? label['org.matrix.msc1767.text'];
  if (text is List && text.isNotEmpty && text.first is Map) {
    final body = (text.first as Map)['body'];
    if (body is String) return body;
  }
  if (text is String) return text;
  return '';
}

Msc4139Prompts? parseMsc4139Prompts(Map<String, dynamic> content) {
  final raw = content[msc4139PromptsKey] ?? content['m.prompts'];
  if (raw is! Map) return null;
  final promptsRaw = raw['prompts'];
  if (promptsRaw is! List) return null;
  final prompts = <Msc4139Prompt>[];
  for (final item in promptsRaw) {
    if (item is! Map) continue;
    final id = item['id'] as String? ?? '';
    if (id.isEmpty) continue;
    prompts.add(
      Msc4139Prompt(
        id: id,
        type: item['type'] == 'input' ? 'input' : 'preset',
        label: _promptLabel(item['label']).isEmpty ? id : _promptLabel(item['label']),
        validator: item['validator'] as String?,
      ),
    );
  }
  if (prompts.isEmpty) return null;
  final introBlock = raw['intro'];
  final intro = introBlock is Map ? _promptLabel(introBlock['content']) : introBlock as String?;
  final scope = (raw['scope'] as List?)?.whereType<String>().toList();
  return Msc4139Prompts(prompts: prompts, intro: intro, scope: scope);
}

Map<String, Object?> conversationReplyContent({
  required String promptId,
  required String label,
  required String rootEventId,
}) {
  return {
    msc4139UsedPromptKey: {'id': promptId},
    'org.matrix.msc1767.text': label,
    'body': label,
    'msgtype': 'm.text',
    'm.relates_to': threadRelation(rootEventId),
    'm.in_reply_to': {'event_id': rootEventId, 'rel_type': 'm.thread'},
  };
}

Map<String, Object?> rtcDeclineContent({String? notificationEventId}) => {
      if (notificationEventId != null)
        'm.relates_to': {
          'rel_type': 'm.reference',
          'event_id': notificationEventId,
        },
    };

String threadSubscriptionPath(String roomId, String rootId) =>
    '/_matrix/client/unstable/org.matrix.msc4306/rooms/${Uri.encodeComponent(roomId)}/thread/${Uri.encodeComponent(rootId)}/subscription';

const profileAboutKey = 'com.highlife.about';

String? parseProfileAbout(Map<String, dynamic> profile) {
  final value = profile[profileAboutKey];
  if (value is String && value.trim().isNotEmpty) return value;
  return null;
}

final _httpUrl = RegExp(r'https?://[^\s<>"]+', caseSensitive: false);

String? firstHttpUrl(String body) {
  final match = _httpUrl.firstMatch(body);
  if (match == null) return null;
  return match.group(0)!.replaceFirst(RegExp(r'[),.;]+$'), '');
}

class UrlPreview {
  const UrlPreview({
    required this.url,
    this.title,
    this.description,
    this.image,
  });

  final String url;
  final String? title;
  final String? description;
  final String? image;
}

UrlPreview? parseUrlPreview(Map<String, dynamic> payload, String fallbackUrl) {
  final title = payload['og:title'] as String?;
  final description = payload['og:description'] as String?;
  final image = payload['og:image'] as String?;
  final url = payload['og:url'] as String? ?? fallbackUrl;
  if ((title == null || title.isEmpty) &&
      (description == null || description.isEmpty) &&
      (image == null || image.isEmpty)) {
    return null;
  }
  return UrlPreview(
    url: url,
    title: title,
    description: description,
    image: image,
  );
}

bool slidingSyncSupported(Map<String, dynamic>? unstable) {
  if (unstable == null) return false;
  return unstable['org.matrix.simplified_msc3575'] == true ||
      unstable['org.matrix.msc4186'] == true ||
      unstable['org.matrix.msc3575'] == true;
}

class MscRoomSummary {
  const MscRoomSummary({
    required this.roomId,
    this.name,
    this.topic,
    this.avatarUrl,
    this.joinRule,
    this.numJoinedMembers,
  });

  final String roomId;
  final String? name;
  final String? topic;
  final String? avatarUrl;
  final String? joinRule;
  final int? numJoinedMembers;
}

MscRoomSummary parseRoomSummary(Map<String, dynamic> payload, String fallbackId) {
  return MscRoomSummary(
    roomId: payload['room_id'] as String? ?? fallbackId,
    name: payload['name'] as String?,
    topic: payload['topic'] as String?,
    avatarUrl: payload['avatar_url'] as String?,
    joinRule: payload['join_rule'] as String?,
    numJoinedMembers: switch (payload['num_joined_members']) {
      final int value => value,
      final num value => value.toInt(),
      _ => null,
    },
  );
}

bool isKnockJoinRule(String? joinRule) =>
    joinRule == 'knock' || joinRule == 'knock_restricted';

/// famedly `Client.request` prefixes `_matrix`, so strip it from spec paths.
String matrixApiAction(String path) {
  const prefix = '/_matrix';
  if (path.startsWith(prefix)) return path.substring(prefix.length);
  return path.startsWith('/') ? path : '/$path';
}

class SpecCommand {
  const SpecCommand({
    required this.name,
    this.aliases = const [],
    this.description,
    this.args = const [],
  });

  final String name;
  final List<String> aliases;
  final String? description;
  final List<String> args;
}

List<SpecCommand> parseCommandsState(Map<String, dynamic> content) {
  final nested = content[msc4332CommandsState];
  final source =
      nested is Map ? Map<String, dynamic>.from(nested) : content;
  final raw = source['commands'];
  if (raw is! List) return const [];
  final out = <SpecCommand>[];
  for (final item in raw) {
    if (item is! Map) continue;
    final row = Map<String, dynamic>.from(item);
    final name = row['name'] as String? ?? '';
    if (name.isEmpty) continue;
    final aliases =
        (row['aliases'] as List?)?.whereType<String>().toList() ?? const [];
    var args = const <String>[];
    final rawArgs = row['args'];
    if (rawArgs is List) {
      args = rawArgs.whereType<String>().toList();
    } else if (rawArgs is String && rawArgs.isNotEmpty) {
      args = [rawArgs];
    }
    out.add(
      SpecCommand(
        name: name,
        aliases: aliases,
        description: row['description'] as String?,
        args: args,
      ),
    );
  }
  return out;
}

Map<String, Object?> defaultSlidingSyncRequest() => {
      'lists': {
        'all': {
          'ranges': [
            [0, 99],
          ],
          'sort': ['by_notification_level', 'by_recency'],
          'timeline_limit': 1,
        },
      },
    };

List<String> parseSlidingSyncRoomOrder(Map<String, dynamic> payload) {
  final lists = payload['lists'];
  if (lists is Map) {
    final all = lists['all'];
    if (all is Map) {
      final ops = all['ops'];
      if (ops is List) {
        final ids = <String>[];
        for (final op in ops) {
          if (op is! Map) continue;
          final roomIds = op['room_ids'];
          if (roomIds is List) {
            ids.addAll(roomIds.whereType<String>());
          }
        }
        if (ids.isNotEmpty) return ids;
      }
    }
  }
  final rooms = payload['rooms'];
  if (rooms is Map) {
    return rooms.keys.map((key) => key.toString()).toList();
  }
  return const [];
}

String matrixLoginQrPayload({
  required String homeserver,
  String? deviceId,
  String? userId,
}) {
  return Uri(
    scheme: 'matrix',
    host: 'login',
    queryParameters: {
      'hs_url': homeserver,
      if (deviceId != null && deviceId.isNotEmpty) 'device': deviceId,
      if (userId != null && userId.isNotEmpty) 'u': userId,
    },
  ).toString();
}

File: apps/web/src/matrix/specFeatures.ts
/** Wire helpers for Element-parity MSCs. Keep this file free of the live client. */

export const MSC4139_PROMPTS = "org.matrix.msc4139.prompts";
export const MSC4139_USED_PROMPT = "org.matrix.msc4139.used_prompt";
export const MSC4139_REPLY = "org.matrix.msc4139.conversation.reply";
export const MSC4332_COMMANDS = "org.matrix.msc4332.commands";
export const MSC4310_DECLINE = "m.rtc.decline";
export const MSC4310_DECLINE_UNSTABLE = "org.matrix.msc4310.rtc.decline";
export const MSC2545_PACK_STATE = "im.ponies.room_emotes";
export const MSC2545_USER_EMOTES = "im.ponies.user_emotes";
export const MSC4306_SUBSCRIPTION_PREFIX = "/_matrix/client/unstable/org.matrix.msc4306/rooms";
export const MSC3266_SUMMARY = "/_matrix/client/v1/room_summary";
export const MSC3266_SUMMARY_UNSTABLE = "/_matrix/client/unstable/im.nheko.summary/summary";
export const MSC4039_UPLOAD = "org.matrix.msc4039.upload_file";
export const MSC4039_DOWNLOAD = "org.matrix.msc4039.download_file";
export const STICKER_EVENT = "m.sticker";

export type IntentionalMentions = {
  user_ids: string[];
  room?: boolean;
};

export function intentionalMentions(
  body: string,
  memberIds: readonly string[],
): IntentionalMentions {
  const user_ids = [...new Set(memberIds.filter((id) => id && body.includes(id)))];
  const room = /(^|\s)@room\b/i.test(body);
  return room ? { user_ids, room: true } : { user_ids };
}

export function attachMentions(
  content: Record<string, unknown>,
  body: string,
  memberIds: readonly string[],
): Record<string, unknown> {
  return { ...content, "m.mentions": intentionalMentions(body, memberIds) };
}

export type ThreadRelation = {
  rel_type: "m.thread";
  event_id: string;
  is_falling_back: boolean;
  "m.in_reply_to": { event_id: string };
};

export function threadRelation(
  rootId: string,
  replyToId = rootId,
  fallback?: boolean,
): ThreadRelation {
  return {
    rel_type: "m.thread",
    event_id: rootId,
    is_falling_back: fallback ?? replyToId === rootId,
    "m.in_reply_to": { event_id: replyToId },
  };
}

export function threadRootId(content: Record<string, unknown>): string | undefined {
  const rel = content["m.relates_to"];
  if (!rel || typeof rel !== "object" || Array.isArray(rel)) return undefined;
  const relation = rel as Record<string, unknown>;
  if (relation.rel_type !== "m.thread") return undefined;
  return typeof relation.event_id === "string" ? relation.event_id : undefined;
}

export function isThreadFallback(content: Record<string, unknown>): boolean {
  const rel = content["m.relates_to"];
  if (!rel || typeof rel !== "object" || Array.isArray(rel)) return false;
  return (rel as { is_falling_back?: boolean }).is_falling_back === true;
}

/** Main timeline keeps roots and fallback previews; in-thread replies stay in the panel. */
export function belongsOnMainTimeline(content: Record<string, unknown>): boolean {
  const root = threadRootId(content);
  if (!root) return true;
  return isThreadFallback(content);
}

export type GeoPoint = { lat: number; lon: number; description?: string; geoUri: string };

export function parseGeoUri(value: string): { lat: number; lon: number } | null {
  const match = value.trim().match(/^geo:(-?\d+(?:\.\d+)?),(-?\d+(?:\.\d+)?)(?:;|$)/i);
  if (!match) return null;
  const lat = Number(match[1]);
  const lon = Number(match[2]);
  if (!Number.isFinite(lat) || !Number.isFinite(lon)) return null;
  return { lat, lon };
}

export function locationContent(lat: number, lon: number, description?: string): Record<string, unknown> {
  const geoUri = `geo:${lat},${lon}`;
  const body = description?.trim() || geoUri;
  return {
    msgtype: "m.location",
    body,
    geo_uri: geoUri,
    "org.matrix.msc3488.location": { uri: geoUri, ...(description ? { description } : {}) },
    "org.matrix.msc3488.ts": Date.now(),
    "org.matrix.msc1767.text": body,
  };
}

export function parseLocationContent(content: Record<string, unknown>): GeoPoint | null {
  const nested = content["org.matrix.msc3488.location"];
  const uri =
    (nested && typeof nested === "object" && !Array.isArray(nested)
      ? (nested as { uri?: unknown }).uri
      : undefined) ?? content.geo_uri;
  if (typeof uri !== "string") return null;
  const parsed = parseGeoUri(uri);
  if (!parsed) return null;
  const description =
    (nested && typeof nested === "object" && !Array.isArray(nested)
      ? (nested as { description?: unknown }).description
      : undefined) ?? content.body;
  return {
    ...parsed,
    geoUri: uri,
    ...(typeof description === "string" ? { description } : {}),
  };
}

export function openStreetMapUrl(lat: number, lon: number): string {
  return `https://www.openstreetmap.org/?mlat=${lat}&mlon=${lon}#map=16/${lat}/${lon}`;
}

export type ImagePackItem = {
  shortcode: string;
  url: string;
  body: string;
  usage: string[];
};

export function parseImagePack(content: Record<string, unknown>): ImagePackItem[] {
  const images = content.images;
  if (!images || typeof images !== "object" || Array.isArray(images)) return [];
  const out: ImagePackItem[] = [];
  for (const [shortcode, raw] of Object.entries(images as Record<string, unknown>)) {
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) continue;
    const item = raw as Record<string, unknown>;
    const url = typeof item.url === "string" ? item.url : undefined;
    if (!url?.startsWith("mxc://")) continue;
    const usage = Array.isArray(item.usage)
      ? item.usage.filter((value): value is string => typeof value === "string")
      : ["sticker", "emoticon"];
    out.push({
      shortcode,
      url,
      body: typeof item.body === "string" ? item.body : shortcode,
      usage,
    });
  }
  return out;
}

export function stickerContent(item: ImagePackItem): Record<string, unknown> {
  return {
    body: item.body,
    url: item.url,
    info: { mimetype: "image/png" },
  };
}

export type AdvertisedCommand = {
  name: string;
  aliases: string[];
  description?: string;
  args?: string[];
};

export function parseCommandsState(content: Record<string, unknown>): AdvertisedCommand[] {
  const nested = content[MSC4332_COMMANDS];
  const source =
    nested && typeof nested === "object" && !Array.isArray(nested)
      ? (nested as Record<string, unknown>)
      : content;
  const raw = source.commands;
  if (!Array.isArray(raw)) return [];
  const out: AdvertisedCommand[] = [];
  for (const item of raw) {
    if (!item || typeof item !== "object" || Array.isArray(item)) continue;
    const row = item as Record<string, unknown>;
    const name = typeof row.name === "string" ? row.name : "";
    if (!name) continue;
    const aliases = Array.isArray(row.aliases)
      ? row.aliases.filter((value): value is string => typeof value === "string")
      : [];
    const args = Array.isArray(row.args)
      ? row.args.filter((value): value is string => typeof value === "string")
      : undefined;
    out.push({
      name,
      aliases,
      ...(typeof row.description === "string" ? { description: row.description } : {}),
      ...(args ? { args } : {}),
    });
  }
  return out;
}

export function filterCommandSuggestions(
  commands: readonly AdvertisedCommand[],
  typed: string,
): AdvertisedCommand[] {
  const t = typed.trim();
  if (!t.startsWith("/") && !t.startsWith("!")) return [];
  const q = t.slice(1).toLowerCase();
  return commands
    .filter(
      (command) =>
        command.name.toLowerCase().startsWith(q) ||
        command.aliases.some((alias) => alias.toLowerCase().startsWith(q)),
    )
    .slice(0, 8);
}

export function completeCommand(command: AdvertisedCommand, typed: string): string {
  const prefix = typed.trimStart().startsWith("!") ? "!" : "/";
  return `${prefix}${command.name} `;
}

export type Msc4139Prompt = {
  id: string;
  type: "preset" | "input";
  label: string;
  validator?: string;
};

export type Msc4139Prompts = {
  intro?: string;
  scope?: string[];
  prompts: Msc4139Prompt[];
};

function promptLabel(label: unknown): string {
  if (typeof label === "string") return label;
  if (!label || typeof label !== "object" || Array.isArray(label)) return "";
  const text = (label as { "org.matrix.msc1767.text"?: unknown; "m.text"?: unknown })["m.text"]
    ?? (label as { "org.matrix.msc1767.text"?: unknown })["org.matrix.msc1767.text"];
  if (Array.isArray(text) && text[0] && typeof text[0] === "object") {
    const body = (text[0] as { body?: unknown }).body;
    if (typeof body === "string") return body;
  }
  if (typeof text === "string") return text;
  return "";
}

export function parseMsc4139Prompts(content: Record<string, unknown>): Msc4139Prompts | null {
  const raw = content[MSC4139_PROMPTS] ?? content["m.prompts"];
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) return null;
  const block = raw as Record<string, unknown>;
  const promptsRaw = block.prompts;
  if (!Array.isArray(promptsRaw)) return null;
  const prompts: Msc4139Prompt[] = [];
  for (const item of promptsRaw) {
    if (!item || typeof item !== "object" || Array.isArray(item)) continue;
    const row = item as Record<string, unknown>;
    const id = typeof row.id === "string" ? row.id : "";
    if (!id) continue;
    prompts.push({
      id,
      type: row.type === "input" ? "input" : "preset",
      label: promptLabel(row.label) || id,
      ...(typeof row.validator === "string" ? { validator: row.validator } : {}),
    });
  }
  if (prompts.length === 0) return null;
  const intro = promptLabel((block.intro as { content?: unknown } | undefined)?.content)
    || (typeof block.intro === "string" ? block.intro : undefined);
  const scope = Array.isArray(block.scope)
    ? block.scope.filter((value): value is string => typeof value === "string")
    : undefined;
  return { prompts, ...(intro ? { intro } : {}), ...(scope ? { scope } : {}) };
}

export function conversationReplyContent(promptId: string, label: string, rootEventId: string): Record<string, unknown> {
  return {
    [MSC4139_USED_PROMPT]: { id: promptId },
    "org.matrix.msc1767.text": label,
    body: label,
    msgtype: "m.text",
    "m.relates_to": threadRelation(rootEventId),
    "m.in_reply_to": { event_id: rootEventId, rel_type: "m.thread" },
  };
}

export function rtcDeclineContent(notificationEventId?: string): Record<string, unknown> {
  return {
    ...(notificationEventId ? { "m.relates_to": { rel_type: "m.reference", event_id: notificationEventId } } : {}),
  };
}

export function threadSubscriptionPath(roomId: string, threadRootId: string): string {
  return `${MSC4306_SUBSCRIPTION_PREFIX}/${encodeURIComponent(roomId)}/thread/${encodeURIComponent(threadRootId)}/subscription`;
}

export const PROFILE_ABOUT_KEY = "com.highlife.about";

export function parseProfileAbout(profile: Record<string, unknown>): string | undefined {
  const value = profile[PROFILE_ABOUT_KEY];
  return typeof value === "string" && value.trim() ? value : undefined;
}

const HTTP_URL = /\bhttps?:\/\/[^\s<>"']+/i;

export function firstHttpUrl(body: string): string | undefined {
  const match = body.match(HTTP_URL);
  if (!match) return undefined;
  return match[0].replace(/[),.;]+$/, "");
}

export type UrlPreview = {
  url: string;
  title?: string;
  description?: string;
  image?: string;
};

export function parseUrlPreview(payload: Record<string, unknown>, fallbackUrl: string): UrlPreview | null {
  const title = typeof payload["og:title"] === "string" ? payload["og:title"] : undefined;
  const description = typeof payload["og:description"] === "string" ? payload["og:description"] : undefined;
  const image = typeof payload["og:image"] === "string" ? payload["og:image"] : undefined;
  const rawUrl = typeof payload["og:url"] === "string" ? payload["og:url"] : fallbackUrl;
  const url = safeHttpUrl(rawUrl) ?? safeHttpUrl(fallbackUrl);
  if (!url) return null;
  if (!title && !description && !image) return null;
  return { url, ...(title ? { title } : {}), ...(description ? { description } : {}), ...(image ? { image } : {}) };
}

function safeHttpUrl(value: string): string | null {
  try {
    const parsed = new URL(value);
    if (parsed.protocol === "https:" || parsed.protocol === "http:") return parsed.toString();
  } catch {
    /* ignore */
  }
  return null;
}

export function urlPreviewCapabilityEnabled(capabilities: Record<string, unknown> | undefined): boolean {
  if (!capabilities) return false;
  const keys = ["org.matrix.msc4452", "m.url_preview"] as const;
  for (const key of keys) {
    const value = capabilities[key];
    if (value && typeof value === "object" && !Array.isArray(value) && (value as { enabled?: unknown }).enabled === true) {
      return true;
    }
  }
  return false;
}

export function slidingSyncSupported(unstable: Record<string, unknown> | undefined): boolean {
  if (!unstable) return false;
  return unstable["org.matrix.simplified_msc3575"] === true
    || unstable["org.matrix.msc4186"] === true
    || unstable["org.matrix.msc3575"] === true;
}

/**
 * Room state Sliding Sync should keep. `["*", "*"]` dumps every membership
 * change and makes Megolm rotate on historical leaves.
 */
export const SLIDING_SYNC_REQUIRED_STATE: string[][] = [
  ["m.room.create", ""],
  ["m.room.name", ""],
  ["m.room.avatar", ""],
  ["m.room.topic", ""],
  ["m.room.canonical_alias", ""],
  ["m.room.encryption", ""],
  ["m.room.power_levels", ""],
  ["m.room.join_rules", ""],
  ["m.room.guest_access", ""],
  ["m.room.history_visibility", ""],
  ["m.room.tombstone", ""],
  ["m.room.member", "$LAZY"],
  ["m.room.member", "$ME"],
  ["m.space.child", "*"],
  ["m.space.parent", "*"],
  ["org.matrix.msc3401.call", "*"],
  ["org.matrix.msc3401.call.member", "*"],
  ["dev.aiomatrix.host", "*"],
  ["im.vector.modular.widgets", "*"],
  [MSC2545_PACK_STATE, "*"],
];

export function slidingSyncRoomSubscription(): {
  timeline_limit: number;
  required_state: string[][];
} {
  return { timeline_limit: 50, required_state: SLIDING_SYNC_REQUIRED_STATE };
}

export function defaultSlidingLists(): Map<string, {
  ranges: number[][];
  sort: string[];
  timeline_limit: number;
  required_state: string[][];
}> {
  return new Map([
    ["all", {
      ranges: [[0, 99]],
      sort: ["by_notification_level", "by_recency"],
      ...slidingSyncRoomSubscription(),
    }],
  ]);
}

export type RoomSummary = {
  roomId: string;
  name?: string;
  topic?: string;
  avatarUrl?: string;
  joinRule?: string;
  guestCanJoin?: boolean;
  numJoinedMembers?: number;
  worldReadable?: boolean;
};

export function parseRoomSummary(payload: Record<string, unknown>, fallbackId: string): RoomSummary {
  return {
    roomId: typeof payload.room_id === "string" ? payload.room_id : fallbackId,
    ...(typeof payload.name === "string" ? { name: payload.name } : {}),
    ...(typeof payload.topic === "string" ? { topic: payload.topic } : {}),
    ...(typeof payload.avatar_url === "string" ? { avatarUrl: payload.avatar_url } : {}),
    ...(typeof payload.join_rule === "string" ? { joinRule: payload.join_rule } : {}),
    ...(typeof payload.guest_can_join === "boolean" ? { guestCanJoin: payload.guest_can_join } : {}),
    ...(typeof payload.num_joined_members === "number" ? { numJoinedMembers: payload.num_joined_members } : {}),
    ...(typeof payload.world_readable === "boolean" ? { worldReadable: payload.world_readable } : {}),
  };
}

export function knockMembers(events: Array<{ sender: string; content: Record<string, unknown> }>): string[] {
  return events
    .filter((event) => event.content.membership === "knock")
    .map((event) => event.sender);
}

export const WIDGET_MSC4039_CAPABILITIES = [
  MSC4039_UPLOAD,
  MSC4039_DOWNLOAD,
  "org.matrix.msc2762.send.event:m.sticker",
  "org.matrix.msc2762.receive.event:m.sticker",
] as const;
